crypto/rsa/rsa_gen.c

```diff
@@ -31,19 +31,40 @@ static int rsa_builtin_keygen(RSA *rsa, int bits, BIGNUM *e_value,
  */
 int RSA_generate_key_ex(RSA *rsa, int bits, BIGNUM *e_value, BN_GENCB *cb)
 {
-    if (rsa->meth->rsa_keygen)
+    return RSA_generate_multi_prime_key(rsa, bits, RSA_DEFAULT_PRIME_NUM,
+                                        e_value, cb);
+}
+
+int RSA_generate_multi_prime_key(RSA *rsa, int bits, int primes,
+                                 BIGNUM *e_value, BN_GENCB *cb)
+{
+    /* multi-prime is only supported with the builtin key generation */
```

Comment:
I noticed this, and wonder if there are plans to support multi-prime key generation via methods...  or is this simply an encouragement for others to step in and do it?

Reply:
Hmm, I considered this also and if I understand correctly, the methods are typically used by ENGINEs and ENGINEs are typically used by accelerator hardwares. I don't know if it's common that hardwares support RSA more than 2-prime. So I am also wondering...

Reply:
You never know, it might come...  after all, RFC 8017 is fairly recent.

Reply:
A new method in `RSA_method` is added to provide the ability of generating multi-prime keys for ENGINEs

Reply:
Actually this creates an ambiguity of following character. Let's say there is rsa->meth that implements keygen method, but not multi_prime_keygen. As it stands now caller would be allowed to ask for a multi-prime key, a key would be generated using bultin method, but *then* rsa->meth provider wouldn't know what to do with it. I mean that in described situation it would be more appropriate if attempt to generate multi-prime key would have been rejected. One can resolve this by demanding rsa->meth to provide either keygen or multi_prime_keygen, but never both. In such case RSA_generate_\*_key subroutines would have to check both methods to determine course of action [returning error if caller asked for multi-prime, but rsa->meth provider implements keygen method.]

[On the other hand one can wonder if there should be separate method at all. I mean if rsa->meth implements keygen method, it's likely to rely on private key structure anyway, which will remain opaque to this code. And last part means that we don't even know what's there, was it multi-prime, was it (Infenion ;-) TPM, ... I mean keygen is free to do whatever anyway...]